pkg/operator/controller/route-metrics/metrics.go

```diff
@@ -0,0 +1,58 @@
+package routemetrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	// routeMetricsControllerRoutesPerShard reports the number of routes belonging to each
+	// Shard (IngressController) using the route_metrics_controller_routes_per_shard metric.
+	routeMetricsControllerRoutesPerShard = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "route_metrics_controller_routes_per_shard",
+		Help: "Report the number of routes for shards (ingress controllers).",
+	}, []string{"name"})
+
+	// metricsList is a list of metrics for this package.
+	metricsList = []prometheus.Collector{
+		routeMetricsControllerRoutesPerShard,
+	}
+)
+
+func InitializeRouteMetricsControllerRoutesPerShardMetric(shardName string) {
+	SetRouteMetricsControllerRoutesPerShardMetric(shardName, 0)
+}
+
+func GetRouteMetricsControllerRoutesPerShardMetric(shardName string) float64 {
+	metric := &dto.Metric{}
+	routeMetricsControllerRoutesPerShard.WithLabelValues(shardName).Write(metric)
```

Comment:
If the label doesn't exist, WithLabelValues will panic.  It might be better to use GetMetricWithLabelValues and check for error here and elsewhere.

You also might want to do error checking on the Write.  If it doesn't work then you might dereference a nil pointer below.

Reply:
I'll remove this function as it would not be required. The existing value of the metric would only be changed after all the routes for the shard are counted. Thus the old value is not needed to be fetched.

Reply:
Please use `GetMetricWithLabelValues` over `WithLabelValues` where you can in these functions.